doc/horizontal_autoscaling.md

```diff
@@ -0,0 +1,163 @@
+
+## Solution
+
+We will build
+an [Operator](https://coreos.com/blog/introducing-operators.html) that
+do the autoscaling. To be more precise, we will create
+a
+[custom Kubernetes Resource](https://kubernetes.io/docs/concepts/api-extension/custom-resources/) and
+a
+[custom Kuberentes Controller](https://resources.coreos.com/youtube-coreos-fest-2017/writing-a-custom-controller-extending-the-functionality-of-your-cluster). Kuberentes
+is designed to be flexible so adding custom Resources and custom
+Controllers into our cluster does not require modifying the Kubernetes
+source code.
+
+### Training Job Resource
+
+Just
+like
+[Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) is
+a resource that describes a deployment. We will have a training
+job
+[Custom Resource](https://kubernetes.io/docs/concepts/api-extension/custom-resources/) that
+describes the training job.
+
+A pseudo resource declaration (`training_job.yaml`) is as follows:
+
+```yaml
+apiVersion: cloud.paddlepaddle.org/v1
+kind: TrainingJob
+metadata:
+  name: job-1
+spec:
+  trainer:
+    entrypoint: "python train.py"
+    workspace: "/home/job-1/"
+    min-replica: 3
+    max-replica: 6
+    cpu: "500m"
+    mem: "1Gi"
+    gpu: 1
+  pserver:
+    min-replica: 3
+    max-replica: 3
+    cpu: "500m"
+    mem: "1Gi"
+```
+
+The training job controller will create and continuously scale the
+number of trainers and the number of pservers between the
+corresponding `min-replica` and `max-replica`.
+
+The training job custom resource can be created with: `kubectl create
+-f training_job.yaml`.
+
+The custom resource will only be saved on Kuberentes, we will need a
+custom controller that operate on it.
+
+### Training Job Controller
+
+The training job controller run as a Pod. It has the global view of
+the computation resources. It watches the training job resources and
+schedules and scales the training jobs using the Kuberenetes API.
+
+The pseudo controller declaration (`training_job_controller.yaml`) is
+as follows:
+
+```yaml
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: training-job-controller
+spec:
+  replicas: 1
+  template:
+    metadata:
+      labels:
+        name: training-job-controller
+    spec:
+      containers:
+      - name: training-job-controller
+        image: paddlepaddle/training-job-controller
+```
+
+The training job controller can be started by the cluster
+administrator with command: `kubectl create -f
+training_job_controller.yaml`
+
+## Implementation
+
+### Training Job Resource
+
+The training job resource is a custom resource, there are two ways of
+implementing custom resources:
+
+- [Custom Resource Definition (CRD)](https://kubernetes.io/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/),
+  since Kuberentes v1.7.
+- [Third Party Resource (TPR)](https://kubernetes.io/docs/tasks/access-kubernetes-api/extend-api-third-party-resource/),
```

Comment:
just remove TPR i guess.

Reply:
I need to check if our cluster has plan to upgrade to v1.7 :)

Reply:
Can we just put the core functions to a library and implement both ways? Upgrading production cluster may be hard and risky. Implement both would be easy for cluster maintainers.

Reply:
@typhoonzero Sure, the main logic will depend on a "register resource" interface, and there will be two implementations: CRD and TPR.